### ISSUE TITLE
Refactor socket code

### ISSUE THREAD

**Issue by brianmingus** [brianmingus/sklearn-emergent#2]
Using existing library?

**Comment by grubino** [comment]
I would recommend using a REST/RPC library, and sticking to standard codecs for models.  The Native presentation format can be rendered from/translated to JSON during deserialization.  Using HTTP for transport makes sense because the transfers are organized into discrete call/response payloads, even in the async case.

**Comment by grubino** [comment]
if you want lightweight, try cherrypy:

http://www.cherrypy.org/

**Comment by grubino** [comment]
You get a basic web server in like 6 lines of code, but you can customize the crap out of it.  If you want to go slightly heavier, Cornice might be a good bet:

https://cornice.readthedocs.org/en/latest/

**Comment by grubino** [comment]
Cornice has lots of nice bells/whistles, but requires a bit more code to get up and running.  Cherrypy is probably a better bet for a micro service.  I've heard Flask is good too, but I've never used it:
http://flask.pocoo.org/

**Comment by brianmingus** [comment]
I'm not entirely clear on the recommendation. My understanding is that Cornice can't be used because we aren't using HTTP. Anything but minor changes to emergent are outside the scope of the project...

**Comment by grubino** [comment]
Why aren't you using HTTP?

**Comment by grubino** [comment]
What you're doing is RPC, for which HTTP is a perfectly suited, already existing and well-established tool.

**Comment by grubino** [comment]
emergent seems almost identical to JSON-RPC:
https://en.wikipedia.org/wiki/JSON-RPC

Except that you have Native codec as well as JSON, which is actually an even stronger argument for separating transport and encoding layers.  Just a suggestion though.

**Comment by grubino** [comment]
At the very least, I would suggest modularizing transport, codec and computation.  I'll submit a PR when I have a moment.

**Comment by grubino** [comment]
Oh!!!  Now I get it.  This is a client... sklearn is the service that it's consuming...  So my comments should be transferred over to sklearn.

**Comment by brianmingus** [comment]
This is an emergent client that provides the necessary interfaces to be consumed as an sklearn estimator.

**Comment by grubino** [comment]
Right, when I said sklearn, I meant emergent.  Emergent is dumb for not just using HTTP for their RPC transport.  I submitted a PR to demonstrate how you could apply separation of concerns (Transport/Encoding and Computation) to this client.

**Comment by brianmingus** [comment]
Looks good, I went ahead and merged it. Will test later.

**Comment by grubino** [comment]
👍
